### PR TITLE
Add lazy builder for LazyList

### PR DESCRIPTION
**Pull request by NthPortal**
Make `LazyList.newBuilder()` return a builder which does
not evaluate collections passed as arguments to `addAll`.

Fixes scala/bug#11299.

### CODE REVIEW

File: src/library/scala/collection/immutable/LazyList.scala [hunk 1]
```diff
@@ -1042,7 +1021,82 @@ object LazyList extends SeqFactory[LazyList] {
...
+  private final class WithFilter[A] private[LazyList](lazyList: LazyList[A], p: A => Boolean)
+    extends collection.WithFilter[A, LazyList] {
+    private[this] val filtered = lazyList.filter(p)
+    def map[B](f: A => B): LazyList[B] = filtered.map(f)
+    def flatMap[B](f: A => IterableOnce[B]): LazyList[B] = filtered.flatMap(f)
+    def foreach[U](f: A => U): Unit = filtered.foreach(f)
+    def withFilter(q: A => Boolean): collection.WithFilter[A, LazyList] = new WithFilter(filtered, q)
+  }
+
+  private final class LazyBuilder[A] extends Builder[A, LazyList[A]] {
```
Comment (NthPortal) [inline src/library/scala/collection/immutable/LazyList.scala]: could probably be a `ReusableBuilder`

**Review by julienrf — COMMENTED** [review]
I haven’t carefully reviewed the implementation but I have two questions:

1. Can we benchmark/compare the advantages over using concatenation on `LazyList`?
2. `Builder` is documented [here](https://docs.scala-lang.org/overviews/core/architecture-of-scala-213-collections.html#when-a-strict-evaluation-is-preferable-or-unavoidable) to be used for building strict collections. The fact that we have a lazy builder might add confusion, what do you think? Should we clearly specify in `Builder#addAll` that the added collection might not be evaluated at all?

**Comment by NthPortal** [comment]
@julienrf sorry I haven't gotten around to addressing your comments - I'll try to get around to it within the week.

1. sure, I can try to do that
2. Definitely seems worth documenting

Relatedly, do you think it would be worth exposing the builder type and giving it a `lazyAddOne` method that takes a by-name element, as well as possibly a `lazyAddAll` method that takes a by-name collection? The former I think is especially valuable, since it is difficult to make a single element lazy otherwise (whereas for a collection you can usually use an `Iterator` or `View`).

**Comment by NthPortal** [comment]
I intend to wait on getting this merged until after #7558, so that that PR doesn't need more rebasing.

**Comment by julienrf** [comment]
> Relatedly, do you think it would be worth exposing the builder type and giving it a `lazyAddOne` method that takes a by-name element, as well as possibly a `lazyAddAll` method that takes a by-name collection? The former I think is especially valuable, since it is difficult to make a single element lazy otherwise (whereas for a collection you can usually use an `Iterator` or `View`).

I don’t think it is really important. To me the simplest way to lazily add an element to a `LazyList` is to use the lazy `prepend` operation: `computeElement() #:: myLazyList`.

**Comment by NthPortal** [comment]
@julienrf I ran a benchmark for the way it is, and tried to run one for just using `appended`/`appendedAll`, but it overflowed the stack. I think stack safety, if nothing else, is good enough reason to use this design.

**Comment by julienrf** [comment]
Thanks @NthPortal for your work on this! My opinion is still that having a special builder for lazy lists makes things more complex to understand. Also, its `addOne` operation looks weird because it creates a lazy state although the added element has already been evaluated.
I would rather fix the stack overflow issue in `appended` and `appendedAll`, if that’s possible.

But maybe others have a different opinion? I don’t want to relentlessly dismiss your work on that…

**Comment by NthPortal** [comment]
@julienrf I don't know if it's possible to get around overflowing the stack with repeated calls to append methods with a lazy collection. If someone has an idea, I'm all ears, but to me it feels like just one of the tradeoffs of having each call to an append method return an immutable instance and be lazy.

It is worth noting that most normal constructions of `LazyList` do not call `lazyAppendedAll` nested more than a few times at most. For example, take the following trivial (and silly) code for generating a `LazyList` of natural numbers:

```scala
Welcome to Scala 2.13.0-20190129-003112-02adaaa (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_191).
Type in expressions for evaluation. Or try :help.

scala> def naturalsFrom(n: Int): LazyList[Int] = LazyList.fill(1)(n) #::: naturalsFrom(n + 1)
naturalsFrom: (n: Int)LazyList[Int]

scala> naturalsFrom(0).take(10_000_000).length
res0: Int = 10000000
```

It calls `lazyAppendedAll`, but only does so on the `LazyList` containing the immediately prior value, rather than a `LazyList` containing all prior values. Thus, you can take 10 million of them and safely call `.length` without overflowing the stack. However, if you store a `LazyList` in a `var` and keep reassigning and appending to it (which is not idiomatic Scala or suggested behaviour, it should be noted), that is not the case:

```scala
scala> var naturals: LazyList[Int] = 0 #:: LazyList.empty[Int]
naturals: LazyList[Int] = LazyList(?)

scala> for (i <- 1 to 100_000) naturals = naturals appended i

scala> naturals.head // overflows the stack
```

Basically, I don't think there's a stack safety problem with the append methods; I think that using them on a `var` to build a `LazyList` is in general an anti-pattern and doesn't really need to be fixed.

**Comment by julienrf** [comment]
> ```scala
> scala> var naturals: LazyList[Int] = 0 #:: LazyList.empty[Int]
> naturals: LazyList[Int] = LazyList(?)
> 
> scala> for (i <- 1 to 100_000) naturals = naturals appended i
> 
> scala> naturals.head // overflows the stack
> ```

This looks like a regression. The same code runs fine with `Stream` :(

**Comment by NthPortal** [comment]
> This looks like a regression. The same code runs fine with `Stream` :(

That's because `Stream` has a strict state and head. If you evaluate any more of the `Stream`, it will overflow the stack as well.

```scala
scala> var naturals: Stream[Int] = 0 #:: Stream.empty[Int]
       [elided warnings ...]
naturals: Stream[Int] = Stream(0, ?)

scala> for (i <- 1 to 100_000) naturals = naturals appended i

scala> naturals.head // safe
res1: Int = 0

scala> naturals.tail // overflows the stack
```

**Comment by SethTisue** [comment]
I don't have a strong feeling either way, but we do need to decide. I lean towards merging after:

* squash
* improved Scaladoc reflecting the discussion here

**Comment by Ichoran** [comment]
@NthPortal - Can you squash and add a warning about stack-unsafety of append to docs?

**Comment by NthPortal** [comment]
sure - I'll try to get to it this weekend. this week has been busy for me

**Comment by NthPortal** [comment]
Squashed and added docs to append methods. Please let me know if you think they could be improved. cc @SethTisue @Ichoran @julienrf

**Comment by NthPortal** [comment]
is this merge-ready?

**Comment by szeiger** [comment]
This has sailed under my radar so far, so take this as my first impression. IMHO the current behavior is not a bug and the non-stack-safe appending makes this change undesirable. Builders are meant for strict building. The fact that `appendAll` can be implemented in a lazier way is not a good enough reason to actually do so. If you care about laziness you won't be using a Builder in the first place. I'd rather see Builders behave consistently without suddenly appending parts of the collection lazily or overflowing the stack.

**Comment by NthPortal** [comment]
@szeiger to be clear, under this implementation, appending to the builder repeatedly does *not* overflow the stack.

**Comment by szeiger** [comment]
Oh, I see, the comments are on `LazyList` itself, not on the builder.

**Comment by SethTisue** [comment]
@szeiger so what's your leaning, overall, then?

**Comment by NthPortal** [comment]
If this PR is rejected, I think we should make the following changes:

- deprecate `LazyList.newBuilder` because you're not getting any `LazyList` semantics out of it
- change `LazyList.newBuilder` to be backed by a `ListBuffer` rather than an `ArrayBuffer` so that some of the `Buffer` can be GC'd before the `LazyList` has been entirely forced

**Comment by szeiger** [comment]
I can't think of a good enough argument why a builder for a lazy collection should *not* be lazy. With the stack safety concerns out of the way I think this is good to go.

**Comment by NthPortal** [comment]
rebased to fix merge conflict
